Add auction series (#2183)
near/docs: docs/3.tutorials/auction/6-indexing.md

---
id: indexing-historical-data
title: Indexing historical data
sidebar_label: Indexing historical data
---

import {Github, Language} from "@site/src/components/codetabs"

> TODO: change github to main when merged

Reviewer: remove note

In our frontend, we can easily display the previous bid since it's stored in the contract's state. However, we're unable to see previous bids to the auction. An indexer is used to fetch historical data from the blockchain and store it in a database. Since indexers can take a while to set up and can be expensive to run, we will use a pre-defined API point provided by NEAR Blocks to query an indexer they run that will fetch us the data we need.

---

## NEAR Blocks API key

> NEAR Blocks provides a free tier that allows you to make 6 calls per minute, this will be plenty for our usecase. To get an API key head over to https://dash.nearblocks.io/user/overview and sign up. Once signed go to `API Keys` then click `Add key` and give it whatever name you like. 

Reviewer: ```suggestion
NEAR Blocks provides a free tier that allows you to make 6 calls per minute, which will be plenty for our use case. To get an API key, head over to https://dash.nearblocks.io/user/overview and sign up. Once signed go to `API Keys` then click `Add key` and give it whatever name you like. 
```

We'll create a new file named `.env.local` to store our API key. Swap the API key in the example with your own.

<Language value="bash" language="bash" showSingleFName={true}>
    <Github fname=".env.local" 

>         url="https://github.com/PiVortex/auctions-tutorial/blob/main/frontend/.env.local"

Reviewer: should this be `github.com/near-examples/...`?

Author: Yep I can change now, was waiting to be merged

        />
</Language>

We put the API key in a `.env.local` file so the user cannot access it in the browser and use our key elsewhere. We should also add `.env.local` to our `.gitignore` file so it is not pushed to GitHub. However, for the purposes of this tutorial, it has been omitted.

---

## Calling the API endpoint

NextJS allows us to easily create server-side functions with API routes. We need to make this API call on the server-side rather than the client side so as to not expose our API key. We'll create a new file in src/pages/api named `getBidHistory.js`. Here we'll define our function to get the bid history.

<Language value="javascript" language="javascript" showSingleFName={true}>
    <Github fname="getBidHistory.js" 

>         url="https://github.com/PiVortex/auctions-tutorial/blob/main/frontend/src/pages/api/getBidHistory.js#L1-L11"

Reviewer: change URL?

## Using the API Route

In our main page, we'll define a function to call the API route we just created. This function will be called as soon as the fungible token account ID is set and each time the page timer reaches zero.

<Language value="javascript" language="javascript" showSingleFName={true}>
    <Github fname="getBidHistory.js" 
        url="https://github.com/PiVortex/auctions-tutorial/blob/main/frontend/src/pages/index.js#L114-L116"
        start="116" end="124" />
</Language>

The `pastBids` will then be passed into the `Bid` component to be displayed.

---

You may like to explore NEAR Blocks APIs further to see what other data you can retrieve from the blockchain. You can find the documentation at https://api.nearblocks.io/api-docs/

---

## Conclusion

In this short part of the tutorial, we've added the ability to display the previous 5 valid bids made to the auction contract. In doing this we learned how to interact with the NEAR Blocks APIs to retrieve historical data from the blockchain and how to make server-side calls in NextJS to not expose our API key. Now we have a pretty good frontend that displays all the information we need about the auction contract.

> In the [final part](./7-factory.md) of this tutorial series we'll learn how to deploy a factory contract - a contract that deploys other contracts - to make it easier for anyone to launch a new auction.

Reviewer: ```suggestion
In the [final part](./7-factory.md) of this tutorial series you'll learn how to deploy a factory contract - a contract that deploys other contracts - to make it easier for anyone to launch a new auction.
```
